### PR TITLE
[HIVEMALL-101] Separate optimizer implementation

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

Finalize #14 

## What type of PR is it?

Improvement, Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-101

## How was this patch tested?

- Unit test
- Manual test on EMR

---

Todo:

- [x] Compare the accuracy of `-loss logloss` with the current `logress()` UDTF
- [x] Add document that includes general explanation of SGD, regularization and optimization techniques like [MLlib](http://spark.apache.org/docs/latest/mllib-optimization.html) and [sklearn](http://scikit-learn.org/stable/modules/sgd.html#mathematical-formulation)
- [x] Support `-mini_batch` option in a similar way to [what RegressionBaseUDTF does](https://github.com/takuti/incubator-hivemall/blob/5dc6f4eb5a8d8532201f6706673e2381d47d7e70/core/src/main/java/hivemall/regression/RegressionBaseUDTF.java#L247-L251); accumulate gradients over M samples, and update for mean value
- [x] ~~Save samples to external files as the other UDTFs (see LDA/pLSA UDTFs) and add `-iter` option~~ *This should be the other issue* [\[HIVEMALL-108\]](https://issues.apache.org/jira/browse/HIVEMALL-108)  
- [x] Manage cumulative loss for future `-iter` support
- [x] Anything else for mix-server-related things?

### CODE REVIEW

File: core/src/main/java/hivemall/regression/GeneralRegressionUDTF.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        } catch (Throwable e) {
+            throw new UDFArgumentException(e.getMessage());
+        }
+        OptimizerOptions.propcessOptions(cl, optimizerOptions);
+        return cl;
+    }
+
+    @Override
+    protected final void checkTargetValue(final float target) throws UDFArgumentException {
```
Comment (takuti) [inline core/src/main/java/hivemall/regression/GeneralRegressionUDTF.java]: @maropu This is a regressor which simply predicts real values. Why did you create this method? Values only in [0,1] are allowed...?

Comment (myui) [inline core/src/main/java/hivemall/regression/GeneralRegressionUDTF.java]: @takuti Maybe for logistic regression that is actually a classifier taking 0/1 values. @maropu is not expert of machine learning algorithm.

Comment (takuti) [inline core/src/main/java/hivemall/regression/GeneralRegressionUDTF.java]: @myui Ah, it makes sense since originally the generic regressor used `LossFunctions.logisticLoss(target, predicted);`. Thanks!

**Comment by myui** [comment]
@takuti `checkTargetValue()` is need for loss function, e.g., for logistic loss.

**Comment by takuti** [comment]
Yep, that's why logistic loss is not selectable for now. `checkTargetValue()` will again come back later.

**Comment by myui** [comment]
@takuti better to have this kind of documents.
http://spark.apache.org/docs/latest/mllib-optimization.html
http://scikit-learn.org/stable/modules/sgd.html#mathematical-formulation

BTW refer [1,2] for how Spark/scikit incorporates regularized updates. FYI
[1] https://github.com/apache/spark/blob/f830bb9170f6b853565d9dd30ca7418b93a54fe3/mllib/src/main/scala/org/apache/spark/mllib/optimization/Updater.scala
[2] https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/sgd_fast.pyx#L632

**Comment by takuti** [comment]
^^^ since generic regressor does not accept classification loss (e.g. logloss) just like sklearn, I keep removing `checkTargetValue()` from the `GeneralRegression` class

**Comment by takuti** [comment]
I listed TODOs in the top comment. If you have any other things I need to care, plz let me know.

**Comment by myui** [comment]
@takuti `-iter` support should be another ticket. `-minibatch` support can be within this ticket.

Functional tests to confirm accuracy of `-loss logistic` to existing `logress` is required.

**Comment by takuti** [comment]
I supported `-mini_batch` option for [regressor](https://github.com/takuti/incubator-hivemall/blob/c3b89f8a671a1ccf7a0c19e9f061d61c6e0c2807/core/src/main/java/hivemall/regression/GeneralRegressionUDTF.java#L121-L181) and [classifier](https://github.com/takuti/incubator-hivemall/blob/c3b89f8a671a1ccf7a0c19e9f061d61c6e0c2807/core/src/main/java/hivemall/classifier/GeneralClassifierUDTF.java#L122-L182) (same code). 

The idea is just accumulating `new_weight` obtained from `optimizer.update()`. Once `miniBatchSize` samples are observed, a mean value of the accumulated `new_weight` will be set to a model via `model.setWeight`.

For SGD, it's clearly equivalent to [what RegressorBaseUDTF does](https://github.com/takuti/incubator-hivemall/blob/5dc6f4eb5a8d8532201f6706673e2381d47d7e70/core/src/main/java/hivemall/regression/RegressionBaseUDTF.java#L247-L251). However, I'm a little bit afraid if I can do the same thing for Adagrad, Adam, Adadelta and AdagradRDA. (Currently, doing the same thing for Adagrad, Adam and Adadelta are allowed. By contrast, AdagradRDA + `-mini_batch` option is not supported.)

BTW, practically, I observed that the naive Adagrad + `-mini_batch` implementation seems to work correctly as shown in the next comment:

**Comment by takuti** [comment]
I tested generic classifier and regressor on EMR by using the a9a data.

### Classifier

```
set hivevar:n_samples=16281;
set hivevar:total_steps=32562;
```

#### `logress`

```sql
drop table if exists logress_model;
create table logress_model as
select
 feature,
 avg(weight) as weight
from
 (
  select
     logress(add_bias(features), label, '-total_steps ${total_steps}') as (feature, weight)
     -- logress(add_bias(features), label, '-total_steps ${total_steps} -mini_batch 10') as (feature, weight)
  from
     train_x3
 ) t
group by feature;
```

```sql
WITH test_exploded as (
  select
    rowid,
    label,
    extract_feature(feature) as feature,
    extract_weight(feature) as value
  from
    test LATERAL VIEW explode(add_bias(features)) t AS feature
),
predict as (
  select
    t.rowid,
    sigmoid(sum(m.weight * t.value)) as prob,
    CAST((case when sigmoid(sum(m.weight * t.value)) >= 0.5 then 1.0 else 0.0 end) as FLOAT) as label
  from
    test_exploded t LEFT OUTER JOIN
    logress_model m ON (t.feature = m.feature)
  group by
    t.rowid
),
submit as (
  select
    t.label as actual,
    pd.label as predicted,
    pd.prob as probability
  from
    test t JOIN predict pd
      on (t.rowid = pd.rowid)
)
select count(1) / ${n_samples} from submit
where actual = predicted;
```

#### `train_classifier`

```sql
train_classifier(add_bias(features), label, '-loss logloss -opt SGD -reg no -eta simple -total_steps ${total_steps}') as (feature, weight)
-- train_classifier(add_bias(features), label, '-loss logloss -opt SGD -reg no -eta simple -total_steps ${total_steps} -mini_batch 10') as (feature, weight)
```

Results were completely same:

| | online | mini-batch |
|:--|:--:|:--:|
|`logress`| 0.8414716540753026 | 0.848965051286776 |
|`train_classifier`| 0.8414716540753026 | 0.848965051286776 |

### Regression

Solved the a9a label prediction as a regression problem. 

// Since non-generic Adagrad was designed for logistic loss (i.e. classification), we cannot compare it with generic regressor under the exactly same condition.

#### `train_adagrad_regr` (internally uses logistic loss)

```sql
drop table if exists adagrad_model;
create table adagrad_model as
select
 feature,
 avg(weight) as weight
from
 (
  select
     train_adagrad_regr(features, label) as (feature, weight)
  from
     train_x3
 ) t
group by feature;
```

```sql
WITH test_exploded as (
  select
    rowid,
    label,
    extract_feature(feature) as feature,
    extract_weight(feature) as value
  from
    test LATERAL VIEW explode(add_bias(features)) t AS feature
),
predict as (
  select
    t.rowid,
    sigmoid(sum(m.weight * t.value)) as prob
  from
    test_exploded t LEFT OUTER JOIN
    adagrad_model m ON (t.feature = m.feature)
  group by
    t.rowid
),
submit as (
  select
    t.label as actual,
    pd.prob as probability
  from
    test t JOIN predict pd
      on (t.rowid = pd.rowid)
)
select rmse(probability, actual) from submit;
```

### `train_regression`

```sql
train_regression(features, label, '-loss squaredloss -opt AdaGrad -reg no') as (feature, weight)
-- train_regression(features, label, '-loss squaredloss -opt AdaGrad -reg no -mini_batch 10') as (feature, weight)
```

| | online | mini-batch |
|:--|:--:|:--:|
|`train_adagrad_regr` (logistic loss) | 0.3254586866367811 | -- |
|`train_regression` (squared loss) | 0.3356422627079689 | 0.3348889704327727 |

As I mentioned in the last comment, I'm afraid whether the `-mini_batch` option works correctly for Adagrad. Fortunately, this example showed that the option slightly improved the accuracy of prediction in terms of RMSE.

**Comment by myui** [comment]
@takuti `train()` can return current loss and cumulative loss should be managed for future iteration support, e.g., using 
https://github.com/apache/incubator-hivemall/blob/master/core/src/main/java/hivemall/common/ConversionState.java

**Comment by myui** [comment]
@takuti I guess no mix-server-related issues in this PR. Will review for that though.

**Comment by takuti** [comment]
@myui Almost done basically. Could you review when you get a chance?

One thing I like to discuss here is that `GeneralClassifierUDTF` and `GeneralRegressionUDTF` currently has a lot of duplicated code. Thus, current class structure

- Learner Base 
  - Binary Online Classifier 
    - General Classifier
  - Regression Base
    - General Regression

can be modified to

- Learner Base
  - General Predictor Base
    - General Classifier
    - General Regression

for example. 

If it sounds good for @myui, I will do so. Of course it's not mandatory, so keeping the current duplicated code is no problem.

**Comment by myui** [comment]
@takuti It's preferred to have an abstract class. Please create it.

- hivemall.LearnerBase
  - hivemall.GeneralLeanerBase
     - hivemall.classifier.GeneralClassifier
     - hivemall.regression.GeneralRegression

**Comment by takuti** [comment]
@myui Finished~

**Comment by myui** [comment]
@takuti well done :+1: will review.

**Comment by myui** [comment]
@maropu @takuti merged this so huge patch finally.. Thank you for your contribution!
